sdk/webpubsub-socketio-extension/src/SIO/components/negotiate.ts

```diff
@@ -0,0 +1,122 @@
+
+/**
+ * Returns a Express middleware to handle negotiate request whose url starts with `path`
+ *
+ * @param path - the path of negotiate request. For Example `/negotiate`.
+ * @param io - the Socket.IO server or its pure options
+ * @param configureNegotiateOptions - a customized function which defines how to extract information for negotiation from a HTTP request
+ * @returns
+ */
+export function negotiate(
+  io: SIO.Server | AzureSocketIOOptions | AzureSocketIOCredentialOptions,
+  configureNegotiateOptions: ConfigureNegotiateOptions
+): (req: ExpressRequest, res: ExpressResponse, next: any) => void {
+  debug(`getNegotiateExpressMiddleware`);
+
+  const wpsOptions: AzureSocketIOOptions | AzureSocketIOCredentialOptions =
+    io instanceof SIO.Server ? io[WEB_PUBSUB_OPTIONS_PROPERY_NAME] : io;
+  const serviceClient = getWebPubSubServiceClient(wpsOptions);
+
+  return async (req: ExpressRequest, res: ExpressResponse, next: any): Promise<void> => {
+    try {
+      debug("negotiate, start");
+
+      const negotiateOptions = await configureNegotiateOptions(req);
+      const tokenResponse = await serviceClient.getClientAccessToken(negotiateOptions);
+      const url = new URL(tokenResponse.baseUrl);
+      const message: NegotiateResponse = {
+        endpoint: url.origin,
+        path: url.pathname,
+        token: tokenResponse.token,
+      };
+      writeResponse(res, 200, message);
+      debug("negotiate, finished");
+    } catch (e) {
+      writeResponse(res, 500, { message: "Internal Server Error" });
+      debug(`negotiate, error: ${e.message}`);
+    }
+  };
+}
+
+/**
+ * Get a `ConfigureNegotiateOptions` used by `negotaite` to enable passport authentication.
+ * Using this middleware, the user id of passport will be put into the response for negotiate request as a part of JWT token.
+ * When the Socket.IO client connects to the server, the user id in JWT token will be extracted and the corresponding passport will be restored and `socket.request.user` will be available.
+ *
+ * @param store - the storage for `express-session`
+ * @param assignProperty - the property name of passport object in `socket.request`. Default value is `user`. Reference: https://www.jsdocs.io/package/@types/passport#AuthenticateOptions.assignProperty
+ * @returns
+ */
+export function usePassport(store: Store, assignProperty = "user"): ConfigureNegotiateOptions {
+  if (!assignProperty || assignProperty.length === 0) {
+    throw new Error("Valid assignProperty is required.");
+  }
+
+  return async (req: ExpressRequest) => {
+    const query = parse(req.url || "", true).query;
+    const expirationMinutes = Number(query["expirationMinutes"] ?? 600);
+
+    const sessionId = req.headers.cookie.slice("connect.sid=s%3A".length).split(".")[0];
+
+    const session = JSON.parse(store["sessions"][sessionId]);
+    if (session["passport"]) {
+      const passportUserId = session["passport"][assignProperty];
```

Comment:
As we discussed, if the `negotaite` middleware puts after passport middleware, can we directly read this userId from req rather than read from scratch